### PR TITLE
Mechanism for raw data previews

### PR DESCRIPTION
**Pull request by gvardany**
**Description of work.**

This PR introduces a simple mechanism of raw data previews.
Preview is a stateless named common operation that is relevant to the data for a given technique at the given facility.
Preview also has a type, which tells the caller which kind of visualisation is desired on the returned workspace.
The previews could be defined in the same files with the respective loader.
This lays the ground for the up-coming raw data explorer widget in the workbench.

**To test:**
Add a preview like this (you can do it in the end of a loader), don't forget to import IPreview.h and PreviewManager.h:
```
class IviewPreview : public API::IPreview {
public:
  std::string name() const override { return "InstrumentViewPreview"; }
  std::string facility() const override { return "ILL"; }
  std::string technique() const override { return "SANS"; }
  API::IPreview::PreviewType type() const override {
    return API::IPreview::PreviewType::IVIEW;
  }
};

DECLARE_PREVIEW(IviewPreview)
```
Compile, then play with this:
```
from mantid.api import IPreview, PreviewManager, PreviewType
instance = PreviewManager.Instance()
previews = instance.getPreviews("ILL", "SANS")
print(previews)
preview = instance.getPreview("ILL", "SANS", 'InstrumentViewPreview')
assert(preview.type() == PreviewType.IVIEW)
ws = CreateSampleWorkspace()
ws = preview.view(ws)
```

Part of #29249 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/inc/MantidDataHandling/LoadILLSANS.h [hunk 1]
```diff
@@ -7,6 +7,8 @@
 #pragma once
 
 #include "MantidAPI/IFileLoader.h"
+#include "MantidAPI/IPreview.h"
+#include "MantidAPI/PreviewManager.h"
```
Comment (DannyHindson) [inline Framework/DataHandling/inc/MantidDataHandling/LoadILLSANS.h:11]: Should these be removed? Maybe they were in there just to test the mechanism?

File: Framework/DataHandling/src/LoadILLSANS.cpp [hunk 2]
```diff
@@ -876,6 +876,5 @@ LoadILLSANS::getVariableTimeBinning(const NXEntry &entry,
   }
   return binEdges;
 }
-
```
Comment (DannyHindson) [inline Framework/DataHandling/src/LoadILLSANS.cpp:879]: Please can you put this line back in so it doesn't show up as a change?

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I added the extra lines to the LoadILLSans.h file as suggested. I had to add the Mantid::API:: prefix in front of IPreview in the IviewPreview class declaration to get it to compile but no big deal

When I run the test python code the getPreviews line returns an empty vector. I had a play around and I wonder if the problem is a missing extern declaration as I've said in one of my detail comments.

Also the final line of the test python code where it calls view didn't do anything but maybe that's intentional and it will display something later on when further work is done in this area

**Comment by gvardany** [comment]
Thanks for good catches, I addressed those, can you try again? It's true the view shouldn't really show anything yet. It's the GUI widget that will handle that based on the preview type, but this is out of API. What you can test is to override also the view method to run something like Scale and check that it is run.

**Review by DannyHindson — APPROVED** [review]
Looks good now thanks + I tested modifying the workspace in an overridden preview function and that worked.

I think this work you're doing around DrILL and other UI related work would be of interest to some of the Mantid team here at ISIS. We spend a lot of effort on developing and maintaining UIs for specific instruments. Would you be happy to do a quick demo of some of this at some point? Maybe at one of the cross team meetings?

**Pull request merged**